Cannot publish page, when attachements with spaces are present

A user erroneously added a Powerpoint file to a page, that was previously published by the `asciidoctor-maven-plugin`.
After this, no other publish could be made to this page, because the publish task failed:

```bash
[INFO] --- asciidoctor-maven-plugin:2.1.0:process-asciidoc (output-html) @ documentation ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 7 resources
[INFO] Converted /workspace/source/documentation/src/docs/asciidoc/deployment.adoc
[INFO] Converted /workspace/source/documentation/src/docs/asciidoc/MyService.adoc
[INFO] asciidoctor: DEBUG: Finding 'dot' in attributes
[INFO] asciidoctor: DEBUG: Finding 'dot' in environment
[INFO] asciidoctor: DEBUG: Found '/usr/bin/dot' in environment
[INFO] Converted /workspace/source/documentation/src/docs/index.adoc
[INFO] 
[INFO] --- asciidoctor-maven-plugin:2.1.0:process-asciidoc (generate-pdf-doc) @ documentation ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 7 resources
[INFO] Converted /workspace/source/documentation/src/docs/asciidoc/deployment.adoc
[INFO] Converted /workspace/source/documentation/src/docs/asciidoc/MyService.adoc
[INFO] Converted /workspace/source/documentation/src/docs/index.adoc
[INFO] 
[INFO] --- asciidoc-confluence-publisher-maven-plugin:0.12.0:publish (publish-documentation) @ documentation ---
[ERROR] Publishing to Confluence failed: Illegal character in path at index 82: https://<confluence-server>/rest/api/content/573476778/property/my demo Jun 22.pptx-hash
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Skipping MyParent - MyService
[INFO] This project has been banned from the build due to previous failures.
[INFO] ------------------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for MyParent - MyService 0.1.0-SNAPSHOT:
[INFO] 
[INFO] MyService Documentation ........................ FAILURE [02:10 min]
[INFO] MyService Service .............................. SKIPPED
[INFO] MyParent - MyService .............. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:11 min
[INFO] Finished at: 2021-07-13T15:29:50Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.sahli.asciidoc.confluence.publisher:asciidoc-confluence-publisher-maven-plugin:0.12.0:publish (publish-documentation) on project documentation: Publishing to Confluence failed: Illegal character in path at index 82: https://<confluence-server>/rest/api/content/573476778/property/my demo Jun 22.pptx-hash -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Could not perform mvn clean verify, exit code [1]
```

The attachement can be found via Confluence under the following URL:
```bash
https://<confluence-server>/download/attachments/573476778/my%20demo%20Jun%2022.pptx?api=v2
```

This is how we configured the plugin
```xml
                    <plugin>
                        <groupId>org.sahli.asciidoc.confluence.publisher</groupId>
                        <artifactId>asciidoc-confluence-publisher-maven-plugin</artifactId>
                        <version>0.12.0</version>
                        <configuration>
                            <spaceKey>${confluencePublisher.spaceKey}</spaceKey>
                            <ancestorId>${confluencePublisher.pageAncestorId}</ancestorId>
                            <asciidocRootFolder>${confluencePublisher.asciidocRootFolder}</asciidocRootFolder>
                            <rootConfluenceUrl>${confluencePublisher.rootUrl}</rootConfluenceUrl>
                            <username>${confluencePublisher.user}</username>
                            <password>${confluencePublisher.password}</password>
                            <versionMessage>Version ${project.version}</versionMessage>
                            <pageTitlePrefix xml:space="preserve"/>
                            <pageTitleSuffix xml:space="preserve"> [${project.version}]</pageTitleSuffix>
                            <attributes>
                                <version>${project.version}</version>
                                <someOtherKey>value</someOtherKey>
                            </attributes>
                        </configuration>
                        <executions>
                            <execution>
                                <id>publish-documentation</id>
                                <phase>generate-resources</phase>
                                <goals>
                                    <goal>publish</goal>
                                </goals>
                            </execution>
                        </executions>
                    </plugin>
```

--- a/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestFactory.java
+++ b/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestFactory.java
@@ -156,14 +156,7 @@ class HttpRequestFactory {
         assertMandatoryParameter(isNotBlank(spaceKey), "spaceKey");
         assertMandatoryParameter(isNotBlank(title), "title");
 
-        String encodedTitle;
-        try {
-            encodedTitle = URLEncoder.encode(title, "UTF-8");
-        } catch (UnsupportedEncodingException e) {
-            throw new RuntimeException("Could not encode title", e);
-        }
-
-        String searchQuery = this.confluenceRestApiEndpoint + "/content?spaceKey=" + spaceKey + "&title=" + encodedTitle;
+        String searchQuery = this.confluenceRestApiEndpoint + "/content?spaceKey=" + spaceKey + "&title=" + urlEncode(title);
 
         return new HttpGet(searchQuery);
     }
@@ -261,14 +254,14 @@ class HttpRequestFactory {
         assertMandatoryParameter(isNotBlank(contentId), "contentId");
         assertMandatoryParameter(isNotBlank(key), "key");
 
-        return new HttpGet(this.confluenceRestApiEndpoint + "/content/" + contentId + "/property/" + key + "?expand=value");
+        return new HttpGet(this.confluenceRestApiEndpoint + "/content/" + contentId + "/property/" + urlEncode(key) + "?expand=value");
     }
 
     public HttpDelete deletePropertyByKeyRequest(String contentId, String key) {
         assertMandatoryParameter(isNotBlank(contentId), "contentId");
         assertMandatoryParameter(isNotBlank(key), "key");
 
-        return new HttpDelete(this.confluenceRestApiEndpoint + "/content/" + contentId + "/property/" + key);
+        return new HttpDelete(this.confluenceRestApiEndpoint + "/content/" + contentId + "/property/" + urlEncode(key));
     }
 
     public HttpPost setPropertyByKeyRequest(String contentId, String key, String value) {
@@ -277,7 +270,7 @@ class HttpRequestFactory {
         assertMandatoryParameter(isNotBlank(value), "value");
 
         PropertyPayload propertyPayload = new PropertyPayload();
-        propertyPayload.setKey(key);
+        propertyPayload.setKey(urlEncode(key));
         propertyPayload.setValue(value);
 
         HttpPost postRequest = new HttpPost(this.confluenceRestApiEndpoint + "/content/" + contentId + "/property");
@@ -310,7 +303,7 @@ class HttpRequestFactory {
         assertMandatoryParameter(isNotBlank(contentId), "contentId");
         assertMandatoryParameter(isNotBlank(label), "label");
 
-        return new HttpDelete(this.confluenceRestApiEndpoint + "/content/" + contentId + "/label?name=" + label);
+        return new HttpDelete(this.confluenceRestApiEndpoint + "/content/" + contentId + "/label?name=" + urlEncode(label));
     }
 
     private static HttpPost addPageHttpPost(String confluenceRestApiEndpoint, PagePayload pagePayload) {
@@ -359,6 +352,14 @@ class HttpRequestFactory {
         }
 
         return multipartEntityBuilder.build();
+    }
+
+    private static String urlEncode(String value) {
+        try {
+            return URLEncoder.encode(value, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException("Could not url-encode value '" + value + "'", e);
+        }
     }
 
 

--- a/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestFactoryTest.java
+++ b/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestFactoryTest.java
@@ -647,6 +647,25 @@ public class HttpRequestFactoryTest {
     }
 
     @Test
+    public void setPropertyByKeyRequest_withSpaceInKey_returnsHttpPostRequestWithUrlEncodedKey() throws Exception {
+        // arrange
+        String contentId = "1234";
+        String key = "key with space";
+        String value = "38495fsj98wgh";
+
+        // act
+        HttpPost setPropertyByKeyRequest = this.httpRequestFactory.setPropertyByKeyRequest(contentId, key, value);
+
+        // assert
+        assertThat(setPropertyByKeyRequest.getURI().toString(), is(CONFLUENCE_REST_API_ENDPOINT + "/content/" + contentId + "/property"));
+        assertThat(setPropertyByKeyRequest.getFirstHeader("Content-Type").getValue(), is(APPLICATION_JSON_UTF8));
+
+        String jsonPayload = inputStreamAsString(setPropertyByKeyRequest.getEntity().getContent(), UTF_8);
+        String expectedJsonPayload = fileContent(Paths.get(CLASS_LOCATION, "set-property-by-key-with-space-in-key-request-payload.json").toString(), UTF_8);
+        assertThat(jsonPayload, isSameJsonAs(expectedJsonPayload));
+    }
+
+    @Test
     public void getPropertyByKeyRequest_withValidParameters_returnsHttpGetRequest() {
         // arrange
         String contentId = "1234";
@@ -660,6 +679,19 @@ public class HttpRequestFactoryTest {
     }
 
     @Test
+    public void getPropertyByKeyRequest_withSpaceInKey_returnsHttpGetRequestWithUrlEncodedKey() {
+        // arrange
+        String contentId = "1234";
+        String key = "key with space";
+
+        // act
+        HttpGet getPropertyByKeyRequest = this.httpRequestFactory.getPropertyByKeyRequest(contentId, key);
+
+        // assert
+        assertThat(getPropertyByKeyRequest.getURI().toString(), is(CONFLUENCE_REST_API_ENDPOINT + "/content/" + contentId + "/property/key+with+space?expand=value"));
+    }
+
+    @Test
     public void deletePropertyByKeyRequest_withValidParameters_returnsHttpDeleteRequest() {
         // arrange
         String contentId = "1234";
@@ -670,6 +702,19 @@ public class HttpRequestFactoryTest {
 
         // assert
         assertThat(deletePropertyByKeyRequest.getURI().toString(), is(CONFLUENCE_REST_API_ENDPOINT + "/content/" + contentId + "/property/" + key));
+    }
+
+    @Test
+    public void deletePropertyByKeyRequest_withSpaceInKey_returnsHttpDeleteRequestWithUrlEncodedKey() {
+        // arrange
+        String contentId = "1234";
+        String key = "key with space";
+
+        // act
+        HttpDelete deletePropertyByKeyRequest = this.httpRequestFactory.deletePropertyByKeyRequest(contentId, key);
+
+        // assert
+        assertThat(deletePropertyByKeyRequest.getURI().toString(), is(CONFLUENCE_REST_API_ENDPOINT + "/content/" + contentId + "/property/key+with+space"));
     }
 
     @Test

--- a/asciidoc-confluence-publisher-client/src/test/resources/org/sahli/asciidoc/confluence/publisher/client/http/set-property-by-key-with-space-in-key-request-payload.json
+++ b/asciidoc-confluence-publisher-client/src/test/resources/org/sahli/asciidoc/confluence/publisher/client/http/set-property-by-key-with-space-in-key-request-payload.json
@@ -1,0 +1,4 @@
+{
+  "key": "key+with+space",
+  "value": "38495fsj98wgh"
+}

--- a/asciidoc-confluence-publisher-converter/src/main/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePage.java
+++ b/asciidoc-confluence-publisher-converter/src/main/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePage.java
@@ -27,6 +27,8 @@ import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
 import java.nio.charset.Charset;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -152,7 +154,7 @@ public class AsciidocConfluencePage {
         String content = ASCIIDOCTOR.convert(adocContent, options);
         String postProcessedContent = postProcessContent(content,
                 replaceCrossReferenceTargets(pagePath, userAttributes, pageTitlePostProcessor, sourceEncoding, spaceKey),
-                collectAndReplaceAttachmentFileNames(attachmentCollector),
+                collectAndReplaceAttachmentFileNames(attachmentCollector, sourceEncoding),
                 unescapeCdataHtmlContent()
         );
 
@@ -163,9 +165,9 @@ public class AsciidocConfluencePage {
         return (content) -> replaceAll(content, CDATA_PATTERN, (matchResult) -> unescapeHtml(matchResult.group()));
     }
 
-    private static Function<String, String> collectAndReplaceAttachmentFileNames(Map<String, String> attachmentCollector) {
+    private static Function<String, String> collectAndReplaceAttachmentFileNames(Map<String, String> attachmentCollector, Charset sourceEncoding) {
         return (content) -> replaceAll(content, ATTACHMENT_PATH_PATTERN, (matchResult) -> {
-            String attachmentPath = matchResult.group(1);
+            String attachmentPath = urlDecode(matchResult.group(1), sourceEncoding);
             String attachmentFileName = deriveAttachmentName(attachmentPath);
 
             attachmentCollector.put(attachmentPath, attachmentFileName);
@@ -279,4 +281,13 @@ public class AsciidocConfluencePage {
                 .map(String::trim)
                 .collect(toList());
     }
+
+    private static String urlDecode(String value, Charset encoding) {
+        try {
+            return URLDecoder.decode(value, encoding.name());
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException("Could not url-decode value '" + value + "'", e);
+        }
+    }
+
 }

--- a/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
+++ b/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
@@ -1692,6 +1692,20 @@ public class AsciidocConfluencePageTest {
     }
 
     @Test
+    public void attachments_asciiDocWithAttachmentWithSpaceInName_returnsAttachmentWithPathAndName() {
+        // arrange
+        String adocContent = "link:attachment%20with%20space.txt[]";
+        AsciidocPage asciidocPage = asciidocPage(prependTitle(adocContent));
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage, UTF_8, TEMPLATES_FOLDER, dummyAssetsTargetPath());
+
+        // assert
+        assertThat(asciidocConfluencePage.attachments().size(), is(1));
+        assertThat(asciidocConfluencePage.attachments(), hasEntry("attachment with space.txt", "attachment with space.txt"));
+    }
+
+    @Test
     public void keywords_singleKeyword_returnsSingleKeyword() {
         // arrange
         String adoc = "= Page Title\n"

--- a/asciidoc-confluence-publisher-doc/etc/docs/00-index/11-attachments.adoc
+++ b/asciidoc-confluence-publisher-doc/etc/docs/00-index/11-attachments.adoc
@@ -4,7 +4,7 @@ If an AsciiDoc file references a non-AsciiDoc file (e.g. a text file, PDF, ...) 
 added as an attachment to the corresponding Confluence page. Referenced non-AsciiDoc files are resolved relative to the
 location of the referencing AsciiDoc file.
 
-Attachment links may specify a label.
+Attachment links may specify a label:
 
 [listing]
 ....
@@ -14,7 +14,7 @@ link:../files/attachment.txt[Attachment]
 link:../files/attachment.txt[Attachment]
 
 
-If no label is specified, the attachment filename is used as the label.
+If no label is specified, the attachment filename is used as the label:
 
 [listing]
 ....
@@ -22,3 +22,12 @@ link:../files/attachment.txt[]
 ....
 
 link:../files/attachment.txt[]
+
+If the attachment filename contains special characters such as spaces, the attachment filename has to be URL-encoded:
+
+[listing]
+....
+link:../files/attachment%20with%20space%20in%20filename.txt[Attachment with space in filename]
+....
+
+link:../files/attachment%20with%20space%20in%20filename.txt[Attachment with space in filename]

--- a/asciidoc-confluence-publisher-doc/etc/docs/files/attachment with space in filename.txt
+++ b/asciidoc-confluence-publisher-doc/etc/docs/files/attachment with space in filename.txt
@@ -1,0 +1,1 @@
+attachment content

